### PR TITLE
Migrate IndexMetadata.getCreationVersion to IndexVersion

### PR DESCRIPTION
**Pull request by thecoop**
Change `IndexMetadata.getCreationVersion` to return `IndexVersion`

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/snapshots/RestoreService.java [hunk 1]
```diff
@@ -1283,7 +1283,7 @@ public ClusterState execute(ClusterState currentState) {
                 final IndexId index = indexEntry.getValue();
                 final IndexMetadata originalIndexMetadata = metadata.index(index.getName());
                 repositoriesService.getPreRestoreVersionChecks()
-                    .forEach(check -> check.accept(snapshot, originalIndexMetadata.getCreationVersion()));
+                    .forEach(check -> check.accept(snapshot, originalIndexMetadata.getCreationVersion().toVersion()));
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/snapshots/RestoreService.java:1286]: Repositories will be migrated separately

File: docs/reference/migration/apis/feature-migration.asciidoc [hunk 2]
```diff
@@ -56,92 +56,93 @@ Example response:
...
-      "minimum_index_version" : "{version}",
+      "minimum_index_version" : "8100099",
       "migration_status" : "NO_MIGRATION_NEEDED",
       "indices" : [ ]
     }
   ],
   "migration_status" : "NO_MIGRATION_NEEDED"
 }
 --------------------------------------------------
+// TESTRESPONSE[s/8100099/\\d+/]
```
Comment (thecoop) [inline docs/reference/migration/apis/feature-migration.asciidoc:145]: This is currently not working - I'm not sure what the best way of specifying this doc test is

File: qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/FeatureUpgradeIT.java [hunk 3]
```diff
@@ -92,8 +93,8 @@ public void testGetFeatureUpgradeStatus() throws Exception {
                     .findFirst()
                     .orElse(Collections.emptyMap());
 
-                assertThat(feature.size(), equalTo(4));
-                assertThat(feature.get("minimum_index_version"), equalTo(UPGRADE_FROM_VERSION.toString()));
+                assertThat(feature, aMapWithSize(4));
+                assertThat(feature.get("minimum_index_version"), equalTo(Integer.toString(UPGRADE_FROM_VERSION.id)));
```
Comment (thecoop) [inline qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/FeatureUpgradeIT.java:97]: The upgrade version is specified in a env parameter; we'll need to add index version to the test infrastructure as well I think...

File: docs/reference/migration/apis/feature-migration.asciidoc (hunk 2 above)
Comment (williamrandolph) [reply docs/reference/migration/apis/feature-migration.asciidoc:145]: Something this should work:

```
// TESTRESPONSE[s/"minimum_index_version" : "8100099"/"minimum_index_version" : $body.$_path/]
```

The `{version}` placeholder gets inserted for tests and for official docs, but not in previews, but it's in the `x.y.z` format.

Comment (thecoop) [reply docs/reference/migration/apis/feature-migration.asciidoc]: That did it, thanks!

**Review by rjernst — APPROVED** [review]
Nice uses of Map.merge to simplify counters. LGTM.

File: qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/FeatureUpgradeIT.java (hunk 3 above)
Comment (thecoop) [reply qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/FeatureUpgradeIT.java:97]: #97200

File: server/src/main/java/org/elasticsearch/snapshots/RestoreService.java (hunk 1 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/snapshots/RestoreService.java]: #97226

**Pull request merged**